File: src/main/java/org/featx/jvmon/model/FlowResponse.java
package org.featx.jvmon.model;

import lombok.Data;
import lombok.EqualsAndHashCode;
import org.featx.jvmon.exception.ErrorCode;

import java.util.List;

/**
 * @author Excepts
 * @since 2019/10/12 11:14
 */
@Data
@EqualsAndHashCode(callSuper = true)
public class FlowResponse<R> extends ListResponse<R> {

    private String previous;

    private String next;

    private FlowResponse(int code, String message, List<R> result) {
        this.setCode(code);
        this.setMessage(message);
        this.setResult(result);
    }

    public FlowResponse<R> previous(String previous) {
        this.previous = previous;
        return this;
    }

    public FlowResponse<R> next(String next) {
        this.next = next;
        return this;
    }

    public FlowResponse<R> cursor(String previous, String next) {
        this.previous = previous;
        this.next = next;
        return this;
    }

    public static <R> FlowResponse<R> succeeded(List<R> result)  {
        return new FlowResponse<>(ErrorCode.NO_ERROR, null, result);
    }
}

File: src/main/kotlin/org/featx/jvmon/model/FlowResponse.kt
package org.featx.jvmon.model

import org.featx.jvmon.exception.BusinessError
import org.featx.jvmon.exception.ErrorCode

/**
 * @author Excepts
 * @since 2019/10/12 11:14
 */
class FlowResponse<T> private constructor(code: Int, msg: String?, result: List<T>?) :
        ListResponse<T>(code, msg, result) {

    var previous: String? = null
        set(previous) {
            field = this.previous
        }

    var next: String? = null
        set(next) {
            field = this.next
        }

    companion object {
        fun <R> succeeded(result: List<R>): FlowResponse<R> {
            return FlowResponse(ErrorCode.NO_ERROR, null, result)
        }

        fun <R> occur(businessError: BusinessError): FlowResponse<R> {
            return FlowResponse(businessError.code, businessError.message, null)
        }

        fun <R> occur(businessError: BusinessError, extra: List<R>): FlowResponse<R> {
            return FlowResponse(businessError.code, businessError.message, extra)
        }
    }
}
